Edit Product: add and remove tags from a Product

pmusolino opened the pull request:
Closes #2081 

## Description
This will be the last part for managing Product Tags. In particular, the focus of this PR is on the UI and on the logic behind it. I've to say that making it work exactly like in the WordPress app was a little bit tricky because our models are a little bit different. In the WordPress app, the tags are managed as simple strings, but in our case, there is a `ProductTag` entity, so I needed to create a method `mergeTags:` for merging tag string entities with the existing/fetched/added `ProductTag` entities.
In this PR I updated only the existing unit tests based on changed that I made, but I've not included any test because the PR is big, and I don't want to make it more complex. I'll take care to test some methods, in particular, the one used for the Tag tokenization and for Text & Input Handling in a separate PR.

## Changes
- There was an error on `ProductStore`. The updates, inserts, or prunes of the provided StorageProduct's tags using the provided read-only Product's tags weren't working correctly and sometimes was possible to see a tag multiple times in the database if it was connected to multiple Products.
- Created a new model `ProductTagCreate` which will be used to handle the create errors on Product Tags Batch Create endpoint.
- Updated the `ProductTagListMapper` to use `ProductTagCreate` when needed.
- `Product` now is able to encode tags.
- Created `ProductTagsViewController` which handle all the logic around Product Tags UI
- `ProductFormViewController` now is able to present and handle Product Tags

## Testing
1. Navigate to a product without tags, press "Add more details"  button, and choose tags.
2. Play with tags, trying to type them, and also try to add some tags from the suggestion list.
3. Add also some new tags (not already existing).
3. Press "Save", and make sure the tags are now updated based on the new ones.
3. Press "Update" in the Product Detail, and make sure that the tags remotely are updated correctly.

## GIF
<img src="https://user-images.githubusercontent.com/495617/88420653-82928000-cde7-11ea-9629-45b3cc5ddc1c.gif" width="300"/>


## Screenshot with dark mode enabled. 
(I'm not sure this is the final result that we want @Garance91540 with the dark mode enabled).
<img src="https://user-images.githubusercontent.com/495617/88420625-74dcfa80-cde7-11ea-9ebb-009b5831a80c.png" width=300 />


Update release notes:
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

pmusolino commented on Networking/Networking/Model/Product/ProductTagCreate.swift:
```diff
@@ -0,0 +1,67 @@
+
+/// Represents a ProductTag entity, used during a tag creation.
+///
+public struct ProductTagCreate: Decodable {
```
Good idea. I renamed it `ProductTagFromBatchCreation ` and removed `public` ecb6d141db9bfa56084c4b8e2c1282a05f7fb1c7 👍

pmusolino commented on Networking/Networking/Model/Product/ProductTagCreate.swift:
```diff
@@ -0,0 +1,67 @@
+}
+
+// MARK: - CreateError
+public struct CreateError: Decodable {
```
Done here ecb6d141db9bfa56084c4b8e2c1282a05f7fb1c7 👍

pmusolino commented on Networking/Networking/Model/Product/ProductTagCreate.swift:
```diff
@@ -0,0 +1,67 @@
+///
+public struct ProductTagCreate: Decodable {
+    public let siteID: Int64
+    public let tagID: Int64
```
If the API response returns an error, `tagID` (id) is populated with `0` 👍 

Sample Request:
```
{
	"create": [{
		"name": "batchAdd1"
	}, {
		"name": "batchAdd5"
	}]
}

Sample Response:
"create": [{
			"id": 0,
			"error": {
				"code": "term_exists",
				"message": "A term with the name provided already exists in this taxonomy.",
				"data": {
					"status": 400,
					"resource_id": 197
				}
			}
		},
		{
			"id": 201,
			"name": "batchAdd5",
			"slug": "batchadd5",
			"description": "",
			"count": 0,
			"_links": {
				"self": [{
					"href": "https://awootestshop.mystagingwebsite.com/wp-json/wc/v3/products/tags/201"
				}],
				"collection": [{
					"href": "https://awootestshop.mystagingwebsite.com/wp-json/wc/v3/products/tags"
				}]
			}
		}
	]
```

pmusolino commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Edit Tags/ProductTagsViewController.swift:
```diff
@@ -0,0 +1,558 @@
+
+    @IBOutlet weak var tableView: UITableView!
+    @IBOutlet weak var textView: UITextView!
+    @IBOutlet var separatorView: UIView!
```
👍  5148c8b101952f42a60262e1c3635ed9c2e9339e

pmusolino commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Edit Tags/ProductTagsViewController.swift:
```diff
@@ -0,0 +1,558 @@
+
+    private var product: Product
+
+    private let originalTags: [String]
```
Great suggestion 👍 375999303e09787fc7bdaee0476b96da7267a3fe

pmusolino commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Edit Tags/ProductTagsViewController.swift:
```diff
@@ -0,0 +1,558 @@
+
+// MARK: - Data Sources
+
+private protocol ProductTagsDataSource: UITableViewDataSource {
```
This is a good idea 💯 204a471ef799e02db53fe55d2035ca30c9a52ebe

pmusolino commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Edit Tags/ProductTagsViewController.swift:
```diff
@@ -0,0 +1,558 @@
+                }
+            }
+        }
+        return finalTags
```
> If there are multiple fetched tags of the same lowercase name, would they all be selected?

Yes, and you are right. I did a change to handle this case merging it with your solution: 90e84b352724bcf97771a965ee98dd4e4e906768

pmusolino commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Edit Tags/ProductTagsViewController.swift:
```diff
@@ -0,0 +1,558 @@
+
+private protocol ProductTagsDataSource: UITableViewDataSource {
+    var selectedTags: [String] { get set }
+    var searchQuery: String { get set }
```
It was done in that way because the datasource used in `ProductTagsViewController` is declared as a generic `ProductTagsDataSource`, and when it's a `SuggestionsDataSource` we need to update the `selectedTags` and the `searchQuery`, but I agree with you that in this case, they are superfluous. I did a change here 2d89dd9918922dace3e1f554a9295e8d5cf75838. Can you take a second look at it? ☺️

pmusolino commented on Yosemite/YosemiteTests/Stores/ProductStoreTests.swift:
```diff
@@ -604,7 +604,7 @@ final class ProductStoreTests: XCTestCase {
         XCTAssertEqual(storageProduct1?.toReadOnly(), sampleProductMutated())
         XCTAssertEqual(viewStorage.countObjects(ofType: Storage.Product.self), 2)
-        XCTAssertEqual(viewStorage.countObjects(ofType: Storage.ProductTag.self, matching: NSPredicate(format: "siteID == %lld", sampleSiteID)), 5)
+        XCTAssertEqual(viewStorage.countObjects(ofType: Storage.ProductTag.self, matching: NSPredicate(format: "siteID == %lld", sampleSiteID)), 10)
```
The reason is that initially the tags are stored in the wrong way, in fact, I did some manual tests, and I noticed that for some reason there where some duplicated tags (every parameter was duplicated), but the tag was associated with a different product. Basically, the updates, inserts, or prunes of the provided StorageProduct's tags using the provided read-only Product's tags weren't working correctly, and sometimes was possible to see a tag multiple times in the database if it was connected to multiple Products. Quite strange things happened.

I solved it this PR, in the `ProductStore` updating the `handleProductTags:` method, and now the counts are correct, because the `sampleProduct()` method generates 9 tags:
```
let tag1 = ProductTag(siteID: siteID, tagID: 37, name: "room", slug: "room")
        let tag2 = ProductTag(siteID: siteID, tagID: 38, name: "party room", slug: "party-room")
        let tag3 = ProductTag(siteID: siteID, tagID: 39, name: "30", slug: "30")
        let tag4 = ProductTag(siteID: siteID, tagID: 40, name: "20+", slug: "20")
        let tag5 = ProductTag(siteID: siteID, tagID: 41, name: "meeting room", slug: "meeting-room")
        let tag6 = ProductTag(siteID: siteID, tagID: 42, name: "meetings", slug: "meetings")
        let tag7 = ProductTag(siteID: siteID, tagID: 43, name: "parties", slug: "parties")
        let tag8 = ProductTag(siteID: siteID, tagID: 44, name: "graduation", slug: "graduation")
        let tag9 = ProductTag(siteID: siteID, tagID: 45, name: "birthday party", slug: "birthday-party")
```
and the `sampleProductMutated()` generates 5 tags, but some of these with the same site ID and tag id.

```
        let tag1 = ProductTag(siteID: siteID, tagID: 37, name: "something", slug: "something")
        let tag2 = ProductTag(siteID: siteID, tagID: 38, name: "party room", slug: "party-room")
        let tag3 = ProductTag(siteID: siteID, tagID: 39, name: "3000", slug: "3000")
        let tag4 = ProductTag(siteID: siteID, tagID: 45, name: "birthday party", slug: "birthday-party")
        let tag5 = ProductTag(siteID: siteID, tagID: 95, name: "yep", slug: "yep")
```

After that the `handleProductTags:` is called the second time (when a new `sampleProductMutated()` is added), the final count is 10 (4 tags updated, and one added).

pmusolino commented:
Thanks @jaclync for your review!

> let's remove/disable duplicates in the tags like in core?

Good catch! I updated it here 90e84b352724bcf97771a965ee98dd4e4e906768 👍 

> not super sure why the order of the tags doesn't match?

🤔  this is really strange. I did a test and I have the same behavior. I did some checks:
- the test that I implemented initially `testUpdatingProductResultingMantainingTheSameOrderForTags()` is ✅ 
- I printed the arrays of tags inside `handleProductTags:` in `ProductStore`. The `readOnlyProduct` is maintaining the right order. The `NSOrderedSet(array: storageTags)` is mantaing the right order 🤔 🤔 🤔 
Any idea? I noticed that you also used NSOrderedSet for Product Images, but I don't know if you encountered a similar issue.

pmusolino commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Edit Tags/ProductTagsViewController.swift:
```diff
@@ -0,0 +1,558 @@
+        } else if text == "\n" // return anywhere else
+            {
+                return false
+        }
```
Yes 🐛 I'll separate and test it in another PR for sure👍

jaclync commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Edit Tags/ProductTagsDataSource.swift:
```diff
@@ -0,0 +1,100 @@
+protocol ProductTagsDataSource: UITableViewDataSource {
+}
+
+class LoadingDataSource: NSObject, ProductTagsDataSource {
```
```suggestion
final class LoadingDataSource: NSObject, ProductTagsDataSource {
```

jaclync commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Edit Tags/ProductTagsDataSource.swift:
```diff
@@ -0,0 +1,100 @@
+    }
+}
+
+class FailureDataSource: NSObject, ProductTagsDataSource {
```
```suggestion
final class FailureDataSource: NSObject, ProductTagsDataSource {
```

jaclync commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Edit Tags/ProductTagsDataSource.swift:
```diff
@@ -0,0 +1,100 @@
+    }
+}
+
+class SuggestionsDataSource: NSObject, ProductTagsDataSource {
```
```suggestion
final class SuggestionsDataSource: NSObject, ProductTagsDataSource {
```

jaclync commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Edit Tags/ProductTagsDataSource.swift:
```diff
@@ -0,0 +1,100 @@
+        super.init()
+    }
+
+    var availableSuggestions: [String] {
```
```suggestion
    private var availableSuggestions: [String] {
```

jaclync commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Edit Tags/ProductTagsDataSource.swift:
```diff
@@ -0,0 +1,100 @@
+        return suggestions.filter({ !selectedTags.contains($0) })
+    }
+
+    var matchedSuggestions: [String] {
```
```suggestion
    private var matchedSuggestions: [String] {
```

jaclync commented on Networking/Networking/Model/Product/ProductTagFromBatchCreation.swift:
```diff
@@ -0,0 +1,67 @@
+    /// ProductTagFromBatchCreation initializer.
+    ///
+    init(siteID: Int64,
+                tagID: Int64,
+                error: CreateError?,
+                name: String?,
+                slug: String?) {
```
```suggestion
         tagID: Int64,
         error: CreateError?,
         name: String?,
         slug: String?) {
```

pmusolino commented:
Thank you so much @jaclync for your precious feedbacks and for discovering the `.sorted()` method called on tags! 😃 
I modified all the suggested things and fixed the product tags, that, as you said, now are shown in the correct order.
